### ISSUE TITLE
Substrate Solochain Implementation Questions

### ISSUE THREAD

**Issue by blakebyrnes** [polytope-labs/hyperbridge#339]
Hi team,

Really like what you're doing here. I'm trying to integrate to a solochain and want to make sure I understand what is required. I'm planning to use TokenGateway, and my chain uses Grandpa. I want to eventually have 2 native balances pallets synced to Ethereum erc20 contracts, but I'm starting with just 1 balances pallet (eg, no assets). 

I have some questions that would be helpful to me (and probably others until the docs go through some of these things a little more):

1. Rpc/API - will the hyper bridge (or any remote ismp) call runtime apis/rpc calls to my node(s)? Do I need to configure both?
2. It says a co-processor for ismp is optional, but seems to be required if you use a gateway. Might be worth mentioning in docs. I assume it also needs to be a coprocessor in the same network (eg, paseo/kusama/polkadot)
3. Configuring a co-processor seems to be different for each chain spec (local testnet, tesnet, mainnet), but is a hardcoded constant in the runtime. Should this be a genesis config or extrinsic (like `token_gateway_addresses` so we can hook up a testnet to paseo and a mainnet chain to kusama/polkadot?
4. Consensus Clients - do I configure the consensus clients I want to be able to consume/decode? Or the ones that decode my chain? More specifically, if I want to support TokenGateway, do I need to have an ismp_beacon client and an ismp_babe for the paseo/kusama hyperbridge? Or do I configure ismp_grandpa for my own chain? I can't figure out "which direction" this is wanting me to configure.
5. TokenGateway - when/why will token governor send a request to create a new asset on my chain? If I intend to only use two instances of the balances pallet, is there any way to do this without pulling in the assets pallet too?
6. It's mentioned that there's a mapping/governance step to map the erc6160 token to the final erc20 token. Where do we make that request?

Thanks! Feel free to split any of these off to other tickets if this is too much for one place, but I didn't want to spam the issues log.

**Comment by seunlanlege** [comment]
> Rpc/API - will the hyper bridge (or any remote ismp) call runtime apis/rpc calls to my node(s)? Do I need to configure both?

The tesseract relayer relies on rpc methods provided by the [pallet-ismp-rpc](https://crates.io/crates/pallet-ismp-rpc) crate, so yes you will have to integrate this, alongside the [pallet-ismp-runtime-api](https://crates.io/crates/pallet-ismp-runtime-api) it depends on.

> It says a co-processor for ismp is optional, but seems to be required if you use a gateway. Might be worth mentioning in docs. I assume it also needs to be a coprocessor in the same network (eg, paseo/kusama/polkadot)

TokenGateway is a Hyperbridge application, not an ISMP application. You can always build your own token bridge protocol over ISMP

> Configuring a co-processor seems to be different for each chain spec (local testnet, tesnet, mainnet), but is a hardcoded constant in the runtime. Should this be a genesis config or extrinsic (like token_gateway_addresses so we can hook up a testnet to paseo and a mainnet chain to kusama/polkadot?

You can configure your runtime anyhow you'd like, our sdks don't enforce anything here

> Consensus Clients - do I configure the consensus clients I want to be able to consume/decode? Or the ones that decode my chain? More specifically, if I want to support TokenGateway, do I need to have an ismp_beacon client and an ismp_babe for the paseo/kusama hyperbridge? Or do I configure ismp_grandpa for my own chain? I can't figure out "which direction" this is wanting me to configure.

You only need to configure consensus clients for the chains you want to communicate with, For hyperbridge you only need the grandpa client.

> TokenGateway - when/why will token governor send a request to create a new asset on my chain? If I intend to only use two instances of the balances pallet, is there any way to do this without pulling in the assets pallet too?

This happens when someone wants to create an asset on your chain, it is permissionless as you can see in the runtime. If your runtime cannot process this request, nothing blows up, you runtime simply will not receive that asset, and the sender can always recover it through timeouts

> It's mentioned that there's a mapping/governance step to map the erc6160 token to the final erc20 token. Where do we make that request?

For now a Github Issue will suffice

**Comment by blakebyrnes** [comment]
Thanks @seunlanlege!

 > > Configuring a co-processor seems to be different for each chain spec (local testnet, tesnet, mainnet), but is a hardcoded constant in the runtime. Should this be a genesis config or extrinsic (like token_gateway_addresses so we can hook up a testnet to paseo and a mainnet chain to kusama/polkadot?
> 
> You can configure your runtime anyhow you'd like, our sdks don't enforce anything here
> 
It sounds like you're considering TokenGateway to be a bit of a reference "app" on hyperbridge, rather than an sdk. As far as I know, requiring a runtime constant means you do either have to have a separate runtime built for each network, or you have to use a build flag. You can't supply it in a genesis config (without having our own pallet).

> > Consensus Clients - do I configure the consensus clients I want to be able to consume/decode? Or the ones that decode my chain? More specifically, if I want to support TokenGateway, do I need to have an ismp_beacon client and an ismp_babe for the paseo/kusama hyperbridge? Or do I configure ismp_grandpa for my own chain? I can't figure out "which direction" this is wanting me to configure.
> 
> You only need to configure consensus clients for the chains you want to communicate with, For hyperbridge you only need the grandpa client.

Ok, but If I want to communicate with Ethereum _via_ hyperbridge, are the proofs preprocessed by a relayer so that I only need to read grandpa from hyperbridge? Or said differently, am I trusting hyperbridge/relayer to have already done the verification before messages get sent over to me?


> > TokenGateway - when/why will token governor send a request to create a new asset on my chain? If I intend to only use two instances of the balances pallet, is there any way to do this without pulling in the assets pallet too?
> 
> This happens when someone wants to create an asset on your chain, it is permissionless as you can see in the runtime. If your runtime cannot process this request, nothing blows up, you runtime simply will not receive that asset, and the sender can always recover it through timeouts
> 

Got it. So I can fork TokenGateway and basically just cut out this part if I don't want to support non-native assets?

**Comment by seunlanlege** [comment]
> Ok, but If I want to communicate with Ethereum via hyperbridge, are the proofs preprocessed by a relayer so that I only need to read grandpa from hyperbridge? Or said differently, am I trusting hyperbridge/relayer to have already done the verification before messages get sent over to me?

Proofs are preprocessed by Hyperbridge, since it is the *coprocessor*

> Got it. So I can fork TokenGateway and basically just cut out this part if I don't want to support non-native assets?

You can do anything, really

**Comment by seunlanlege** [comment]
> When would I want someone to be able to hydrate a new asset on the chain without adding it myself first?

I would assume you want your chain to be permissionless, but if not. You can also reject these asset creation requests

**Comment by blakebyrnes** [comment]
> > When would I want someone to be able to hydrate a new asset on the chain without adding it myself first?
> 
> I would assume you want your chain to be permissionless, but if not. You can also reject these asset creation requests

Yeah, by "myself", I mean, the chain governance activates a new asset if it's determined to be useful. Will just assume this is not something for now. Thanks for the answers.
